File: npm/src/reporter.js
'use strict';

const chalk    = require('chalk');
const Table    = require('cli-table3');
const { loadSessions } = require('./storage');
const { durationMinutes } = require('./tracker');

function printReport(days = 30) {
  const sessions  = fetchCompleted(days);

  if (!sessions.length) {
    console.log(chalk.yellow(`\nNo completed sessions in the last ${days} days.\n`));
    console.log(`Run ${chalk.cyan('bugclock start')} to begin tracking.\n`);
    return;
  }

  const stats = computeStats(sessions);

  console.log('\n' + chalk.cyan('━'.repeat(52)));
  console.log(chalk.bold.cyan('  bugclock') + chalk.dim(`  · last ${days} days`));
  console.log(chalk.cyan('━'.repeat(52)));

  // Key metrics
  console.log('');
  printMetrics(stats);

  // Daily chart
  printDailyChart(sessions, Math.min(days, 14));

  // Recent sessions table
  printSessionsTable(sessions.slice(0, 10));

  // Top files
  printTopFiles(sessions);

  const auto = sessions.filter(s => s.source === 'auto').length;
  if (auto) {
    console.log(chalk.dim(`  ${sessions.length - auto} manual · ${auto} auto-inferred from commits\n`));
  }
}

function exportJson(days = 30) {
  return JSON.stringify(
    fetchCompleted(days).map(s => ({ ...s, duration_minutes: durationMinutes(s) })),
    null, 2
  );
}

function exportCsv(days = 30) {
  const sessions = fetchCompleted(days);
  const header   = 'id,started_at,ended_at,duration_minutes,status,files,author,source,notes,commit_hash';
  const rows     = sessions.map(s =>
    [
      s.id,
      s.started_at,
      s.ended_at  || '',
      durationMinutes(s) ?? '',
      s.status,
      `"${(s.files || []).join('; ')}"`,
      s.author    || '',
      s.source    || '',
      `"${(s.notes || '').replace(/"/g, '""')}"`,
      s.commit_hash || '',
    ].join(',')
  );
  return [header, ...rows].join('\n');
}

// ── Private ───────────────────────────────────────────────────────────────────

function fetchCompleted(days) {
  const cutoff = new Date(Date.now() - days * 86400000).toISOString();
  return loadSessions()
    .filter(s => s.status !== 'active' && s.started_at >= cutoff)
    .slice().reverse();
}

function computeStats(sessions) {
  const durations = sessions.map(durationMinutes).filter(d => d != null);
  const resolved  = sessions.filter(s => s.status === 'resolved').length;
  const total     = durations.reduce((a, b) => a + b, 0);
  return {
    total:    sessions.length,
    resolved,
    abandoned: sessions.length - resolved,
    totalMin: Math.round(total * 10) / 10,
    avgMin:   durations.length ? Math.round(total / durations.length * 10) / 10 : 0,
    longestMin: durations.length ? Math.round(Math.max(...durations) * 10) / 10 : 0,
  };
}

function printMetrics(s) {
  const metrics = [
    ['Total Sessions', String(s.total),        chalk.white],
    ['Resolved',       String(s.resolved),     chalk.green],
    ['Abandoned',      String(s.abandoned),    chalk.red],
    ['Total Time',     `${s.totalMin} min`,    chalk.yellow],
    ['Avg / Session',  `${s.avgMin} min`,      chalk.magenta],
    ['Longest',        `${s.longestMin} min`,  chalk.white],
  ];
  metrics.forEach(([label, value, color]) => {
    console.log(`  ${chalk.dim((label + ':').padEnd(18))} ${color(value)}`);
  });
  console.log('');
}

function printDailyChart(sessions, days) {
  const buckets = {};
  for (const s of sessions) {
    const d = durationMinutes(s);
    if (d == null) continue;
    const day = s.started_at.slice(0, 10);
    buckets[day] = (buckets[day] || 0) + d;
  }
  if (!Object.keys(buckets).length) return;

  const maxVal    = Math.max(...Object.values(buckets)) || 1;
  const BAR_WIDTH = 20;
  const today     = new Date();

  console.log(chalk.bold('  Daily Debug Time (minutes)'));
  for (let i = days - 1; i >= 0; i--) {
    const d   = new Date(today);
    d.setDate(today.getDate() - i);
    const day = d.toISOString().slice(0, 10);
    const val = buckets[day] || 0;
    const filled = Math.round((val / maxVal) * BAR_WIDTH);
    const bar = chalk.cyan('█'.repeat(filled)) + chalk.dim('░'.repeat(BAR_WIDTH - filled));
    console.log(`  ${chalk.dim(day)}  ${String(Math.round(val * 10) / 10).padStart(5)}  ${bar}`);
  }
  console.log('');
}

function printSessionsTable(sessions) {
  const table = new Table({
    head: ['#', 'Started', 'Duration', 'Status', 'Files', 'Author'].map(h => chalk.bold(h)),
    colWidths: [5, 18, 10, 12, 32, 16],
    style: { compact: true },
  });

  for (const s of sessions) {
    const dur  = durationMinutes(s);
    const files = (s.files || []).join(', ') || '—';
    const statusColor = { resolved: chalk.green, abandoned: chalk.red }[s.status] || chalk.yellow;
    table.push([
      String(s.id),
      s.started_at.slice(0, 16).replace('T', ' '),
      dur != null ? `${dur} min` : '—',
      statusColor(s.status),
      files.length > 30 ? files.slice(0, 27) + '...' : files,
      s.author || '—',
    ]);
  }

  console.log(chalk.bold('  Recent Sessions'));
  console.log(table.toString());
  console.log('');
}

function printTopFiles(sessions) {
  const timeMap  = {};
  const countMap = {};
  for (const s of sessions) {
    const d = durationMinutes(s) || 0;
    for (const f of s.files || []) {
      timeMap[f]  = (timeMap[f]  || 0) + d;
      countMap[f] = (countMap[f] || 0) + 1;
    }
  }
  if (!Object.keys(timeMap).length) return;

  const top = Object.entries(timeMap)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 6);

  const table = new Table({
    head: ['File', 'Sessions', 'Total Time'].map(h => chalk.bold(h)),
    style: { compact: true },
  });
  for (const [f, t] of top) {
    table.push([chalk.cyan(f), String(countMap[f]), `${Math.round(t * 10) / 10} min`]);
  }

  console.log(chalk.bold('  Most-Debugged Files'));
  console.log(table.toString());
  console.log('');
}

module.exports = { printReport, exportJson, exportCsv };

File: claude_debug_tracker/reporter.py
"""
Rich-powered analytics and reporting.
Produces terminal reports, JSON/CSV exports, and ASCII bar charts.
"""

import json
import csv
import io
from collections import defaultdict
from datetime import datetime, timedelta
from typing import List, Optional

from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich import box
from rich.text import Text
from rich.columns import Columns

from .storage import get_db, init_db
from .tracker import duration_minutes, list_sessions

console = Console()


# ── Public API ────────────────────────────────────────────────────────────────

def print_report(days: int = 30):
    """Print a full analytics report to the terminal."""
    init_db()
    sessions = _fetch_completed(days)

    if not sessions:
        console.print(
            Panel(
                "[yellow]No completed debug sessions in the last "
                f"{days} days.[/yellow]\n\n"
                "Run [bold cyan]bugclock start[/bold cyan] to begin tracking.",
                title="Claude Debug Tracker",
                border_style="cyan",
            )
        )
        return

    stats = _compute_stats(sessions)

    # ── Header ────────────────────────────────────────────────────────────────
    console.print()
    console.print(
        Panel(
            f"[bold cyan]Claude Debug Tracker[/bold cyan]  "
            f"[dim]last {days} days[/dim]",
            border_style="cyan",
            padding=(0, 2),
        )
    )

    # ── Key metrics ───────────────────────────────────────────────────────────
    metric_panels = [
        Panel(
            f"[bold white]{stats['total_sessions']}[/bold white]\n[dim]Total Sessions[/dim]",
            border_style="blue",
        ),
        Panel(
            f"[bold green]{stats['resolved']}[/bold green]\n[dim]Resolved[/dim]",
            border_style="green",
        ),
        Panel(
            f"[bold red]{stats['abandoned']}[/bold red]\n[dim]Abandoned[/dim]",
            border_style="red",
        ),
        Panel(
            f"[bold yellow]{stats['total_minutes']} min[/bold yellow]\n[dim]Total Debug Time[/dim]",
            border_style="yellow",
        ),
        Panel(
            f"[bold magenta]{stats['avg_minutes']} min[/bold magenta]\n[dim]Avg per Session[/dim]",
            border_style="magenta",
        ),
        Panel(
            f"[bold white]{stats['longest_minutes']} min[/bold white]\n[dim]Longest Session[/dim]",
            border_style="white",
        ),
    ]
    console.print(Columns(metric_panels, equal=True, expand=True))
    console.print()

    # ── Daily chart ───────────────────────────────────────────────────────────
    _print_daily_chart(sessions, days=min(days, 14))

    # ── Recent sessions table ─────────────────────────────────────────────────
    _print_sessions_table(sessions[:10])

    # ── Top debugged files ────────────────────────────────────────────────────
    _print_top_files(sessions)

    # ── Auto vs manual breakdown ──────────────────────────────────────────────
    auto = sum(1 for s in sessions if s.get("source") == "auto")
    manual = len(sessions) - auto
    if auto:
        console.print(
            f"[dim]Session sources: {manual} manual · {auto} auto-inferred from commits[/dim]\n"
        )


def export_json(days: int = 30) -> str:
    sessions = _fetch_completed(days)
    enriched = []
    for s in sessions:
        s["duration_minutes"] = duration_minutes(s)
        enriched.append(s)
    return json.dumps(enriched, indent=2)


def export_csv(days: int = 30) -> str:
    sessions = _fetch_completed(days)
    buf = io.StringIO()
    fields = ["id", "started_at", "ended_at", "duration_minutes", "status",
              "files", "author", "source", "notes", "commit_hash"]
    writer = csv.DictWriter(buf, fieldnames=fields, extrasaction="ignore")
    writer.writeheader()
    for s in sessions:
        s["duration_minutes"] = duration_minutes(s)
        s["files"] = "; ".join(s.get("files") or [])
        writer.writerow(s)
    return buf.getvalue()


# ── Private helpers ───────────────────────────────────────────────────────────

def _fetch_completed(days: int) -> List[dict]:
    init_db()
    cutoff = (datetime.utcnow() - timedelta(days=days)).isoformat()
    conn = get_db()
    rows = conn.execute(
        "SELECT * FROM sessions WHERE status != 'active' AND started_at >= ? "
        "ORDER BY started_at DESC",
        (cutoff,),
    ).fetchall()
    conn.close()

    import json as _json
    result = []
    for r in rows:
        d = dict(r)
        d["files"] = _json.loads(d.get("files") or "[]")
        result.append(d)
    return result


def _compute_stats(sessions: List[dict]) -> dict:
    durations = [d for s in sessions if (d := duration_minutes(s)) is not None]
    resolved = sum(1 for s in sessions if s["status"] == "resolved")
    abandoned = len(sessions) - resolved
    total = round(sum(durations), 1)
    avg = round(total / len(durations), 1) if durations else 0.0
    longest = round(max(durations), 1) if durations else 0.0
    return {
        "total_sessions": len(sessions),
        "resolved": resolved,
        "abandoned": abandoned,
        "total_minutes": total,
        "avg_minutes": avg,
        "longest_minutes": longest,
    }


def _print_daily_chart(sessions: List[dict], days: int = 14):
    # Bucket by day
    buckets: dict = defaultdict(float)
    for s in sessions:
        d = duration_minutes(s)
        if d is None:
            continue
        day = s["started_at"][:10]
        buckets[day] += d

    if not buckets:
        return

    today = datetime.utcnow().date()
    day_range = [(today - timedelta(days=i)).isoformat() for i in range(days - 1, -1, -1)]
    max_val = max(buckets.values()) or 1
    bar_width = 20

    table = Table(title="Daily Debug Time (minutes)", box=box.SIMPLE, show_header=True)
    table.add_column("Date", style="dim", width=12)
    table.add_column("Minutes", justify="right", width=8)
    table.add_column("", width=bar_width + 2)

    for day in day_range:
        val = buckets.get(day, 0.0)
        filled = int((val / max_val) * bar_width)
        bar = "[cyan]" + "█" * filled + "[/cyan]" + "░" * (bar_width - filled)
        table.add_row(day, f"{round(val, 1)}", bar)

    console.print(table)
    console.print()


def _print_sessions_table(sessions: List[dict]):
    table = Table(title="Recent Sessions", box=box.ROUNDED, show_lines=False)
    table.add_column("#", style="dim", width=5)
    table.add_column("Started", width=20)
    table.add_column("Duration", justify="right", width=10)
    table.add_column("Status", width=10)
    table.add_column("Files", width=30)
    table.add_column("Author", width=16)

    for s in sessions:
        dur = duration_minutes(s)
        dur_str = f"{dur} min" if dur is not None else "—"
        files = ", ".join(s.get("files") or []) or "[dim]—[/dim]"
        if len(files) > 30:
            files = files[:27] + "..."

        status_color = {"resolved": "green", "abandoned": "red"}.get(s["status"], "yellow")
        table.add_row(
            str(s["id"]),
            s["started_at"][:16].replace("T", " "),
            dur_str,
            f"[{status_color}]{s['status']}[/{status_color}]",
            files,
            s.get("author") or "—",
        )

    console.print(table)
    console.print()


def _print_top_files(sessions: List[dict]):
    file_time: dict = defaultdict(float)
    file_count: dict = defaultdict(int)

    for s in sessions:
        d = duration_minutes(s) or 0.0
        for f in s.get("files") or []:
            file_time[f] += d
            file_count[f] += 1

    if not file_time:
        return

    top = sorted(file_time.items(), key=lambda x: x[1], reverse=True)[:8]

    table = Table(title="Most-Debugged Files", box=box.SIMPLE)
    table.add_column("File", style="cyan")
    table.add_column("Sessions", justify="right", width=10)
    table.add_column("Total Time", justify="right", width=12)

    for filepath, total in top:
        table.add_row(filepath, str(file_count[filepath]), f"{round(total, 1)} min")

    console.print(table)
    console.print()
